clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/deploy/util/id/DcosSpinnakerAppId.groovy

```
@@ -0,0 +1,144 @@
+     *         <p/>
+     *         Ex: {@code foo/bar}
+     * @see #getSafeRegion()
+     */
+    public String getUnsafeRegion() {
+        region
+    }
+
+    /**
+     * @return The "safe" DC/OS region (a.k.a the group in which the marathon application lives). This is returned as a
+     *         relative path, meaning no preceeding underscore. Will never be null.
+     *         <p/>
+     *         Deemed safe because backslashes are replaced with underscores.
+     *         <p/>
+     *         Ex: {@code acct_foo_bar}
+     * @see #getSafeRegion()
+     */
+    public String getSafeRegion() {
+        region.replaceAll(MarathonPathId.PART_SEPARATOR, SAFE_REGION_SEPARATOR)
+    }
+
+    public Names getServerGroupName() {
+        Names.parseName(marathonPath.last().get())
+    }
+
+    public MarathonPathId toMarathonPathId() {
+        marathonPath
+    }
+
+    @Override
+    public String toString() {
+        marathonPath.toString()
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o)
+            return true
+        if (o == null || getClass() != o.getClass())
+            return false
+        def dcosPathId = (DcosSpinnakerAppId) o
+        return dcosPathId.toString() == toString()
+    }
+
+    @Override
+    public int hashCode() {
+        return toString().hashCode()
+    }
+
+    private static void validateServerGroupName(final String serverGroupName) {
+        def service = Names.parseName(serverGroupName)
+
+        checkArgument(!nullToEmpty(service.app).trim().empty, "The server group app should not be null, empty, or blank.")
+        checkArgument(service.sequence >= 0, "The server group sequence should not be negative or null.")
+    }
+
+    public static Optional<DcosSpinnakerAppId> from(String marathonAppId) {
+        try {
+            Optional.of(new DcosSpinnakerAppId(marathonAppId))
+        } catch (IllegalArgumentException e) {
```

I'm not sure I like how we're catching IllegalArgumentExceptions in the Optional factory methods. Typically an IllegalArgumentException is a guard against the method caller making a mistake (i.e. the documentation says you can't pass a null/empty account or name, but the caller didn't properly account for that scenario). Before, I think we were handling that and didn't need an Optional alternative.

---

Before we were allowing invalid objects to be built and just giving them a validate function that told you whether an object is valid or not. I don't think we should be allowing for technically invalid objects to ever be built. Now I could just try and refactor this a bit to either allow us to reuse validation logic in the constructors (or move the try block to a "validate" function).

---

It doesn't look like invalid ids would get created (https://github.com/cerner/clouddriver/pull/10/files#diff-589ac0bedd6bcd7cb554c06ef4e59e27L54). There was a stipulation in the doc that you should ensure the id is valid first (otherwise it would throw an exception, not create the invalid id)

However, I think encapsulating that in the static factory method is good (with the Optional return), though I think only the "parse" factory methods (i.e. the ones that take in a marathonAppId) should have it. Doesn't seem necessary for the ones that take in discreet parts, since all the calling sites should already have valid values for all of those.

---

> It doesn't look like invalid ids would get created

I meant creating the object even within the static factory methods, we should be validating input before creation of an object. It isn't a big issue, but I would prefer we validate before creation. Which we can still do if we just move the checks to static factory methods instead of the constructors (if they are made private).

> Doesn't seem necessary for the ones that take in discreet parts, since all the calling sites should already have valid values for all of those.

Not necessarily, as we have seen on a couple of occasions where that part doesn't follow the marathon pattern (capitals in the account, region, etc for example). Granted the validators for operations may take care of that for us.